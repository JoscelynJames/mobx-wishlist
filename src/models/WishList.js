import { types } from 'mobx-state-tree';

const data = {
    'name': 'African Grey',
    'price': 1500,
    'image': 'https://i.pinimg.com/originals/90/dc/6b/90dc6b867860400b9001327e7c29759f.jpg'
}

export const WishListItem = types.model({
    name: types.string,
    price: types.number,
    // image: types.optional(types.string, '')
    // way defaults to optional
    image: ''
})
.actions(self => ({
    changeName(newName) {
        self.name = newName;
    },
    changePrice(newPrice) {
        self.price = newPrice;
    },
    changeImage(newImage) {
        self.image = newImage;
    }

    // one way to writing actions
    // function changeName(newName) {
    //     self.name = newName;
    // }

    // return {
    //     changeName
    // }

}));

export const WishList = types.model({
    items: types.optional(types.array(WishListItem), []),
})
.actions(self => ({
    add(item) {
        self.items.push(item);
    }
}))
.views(self => ({
    get totalPrice() {
        return self.items.reduce((sum, entry) => sum + entry.price, 0)
    }
}))
