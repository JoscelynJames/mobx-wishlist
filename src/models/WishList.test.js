import { WishList, WishListItem } from './WishList';
import { getSnapshot, onSnapshot, onPatch } from 'mobx-state-tree';
import { reaction } from 'mobx';

it('can create an instance of a model', () => {
    const item = WishListItem.create({
        name: 'African Grey',
        price: 1500
    });

    expect(item.price).toBe(1500);
    expect(item.image).toBe("");
    item.changeName('Cockatoo');
    expect(item.name).toBe('Cockatoo');
    item.changePrice(2000);
    expect(item.price).toBe(2000);

});

it('can create a WishLsit', () => {
    const wishList = WishList.create({
        items: [
            {
                name: 'African Grey',
                price: 1500
            }
        ]
    });

    expect(wishList.items.length).toBe(1);
    expect(wishList.items[0].price).toBe(1500);

});

it('can add new items (onSnapshot)', () => {
    const wishList = WishList.create();
    const states = [];
    onSnapshot(wishList, snapshot => {
        states.push(snapshot)
    });

    wishList.add(
        WishListItem.create({
            name: 'Sun Conure',
            price: 1000
        })
    );

    expect(wishList.items.length).toBe(1);
    expect(wishList.items[0].price).toBe(1000);
    wishList.items[0].changePrice(800);
    expect(wishList.items[0].price).toBe(800);

    expect(getSnapshot(wishList)).toMatchSnapshot();
    expect(states).toMatchSnapshot();

    // jest has snapshots that will make file of the object and compare it to that
    // expect(getSnapshot(wishList)).toEqual({
    //     items: [
    //         {
    //             name: 'Sun Conure',
    //             price: 800,
    //             image: ''
    //         }
    //     ]
    // })

});

it('can add new items (onPatch)', () => {
    const wishList = WishList.create();
    const patches = [];
    onPatch(wishList, snapshot => {
        patches.push(snapshot)
    });

    wishList.add(
        WishListItem.create({
            name: 'Sun Conure',
            price: 1000
        })
    );

    expect(patches).toMatchSnapshot();

});

it('can add new items (onPatch)', () => {
    const wishList = WishList.create({
        items: [
            {
                name: 'Pineapple Conure',
                price: 600
            },
            {
                name: 'Green Cheek Conure',
                price: 600
            }
        ]
    });

    expect(wishList.totalPrice).toBe(1200);

    let changed= 0;
    // reaction will listen to any changes on set value/funtion ( wishList.totalPrice) 
    // and call a callback whenever a change is detected 
    reaction(() => wishList.totalPrice, () => changed++);

    expect(changed).toBe(0);
    wishList.items[0].changeName('Dusky Conure')
    expect(changed).toBe(0);
    wishList.items[0].changePrice(700)
    expect(changed).toBe(1);

});