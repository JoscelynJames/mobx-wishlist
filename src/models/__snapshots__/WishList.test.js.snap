// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add new items (onPatch) 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "image": "",
      "name": "Sun Conure",
      "price": 1000,
    },
  },
]
`;

exports[`can add new items (onPatch) 2`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "image": "",
      "name": "Sun Conure",
      "price": 1000,
    },
  },
]
`;

exports[`can add new items (onSnapshot) 1`] = `
Object {
  "items": Array [
    Object {
      "image": "",
      "name": "Sun Conure",
      "price": 800,
    },
  ],
}
`;

exports[`can add new items (onSnapshot) 2`] = `
Array [
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "Sun Conure",
        "price": 1000,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "Sun Conure",
        "price": 800,
      },
    ],
  },
]
`;
